hail/python/hail/experimental/dnd/array.py

```diff
@@ -0,0 +1,260 @@
+                "name": "StreamBlockBufferSpec"}}})
+
+    @staticmethod
+    def from_matrix_table(
+            mt: MatrixTable,
+            entry_field: str,
+            *,
+            n_partitions: Optional[int] = None,
+            block_size: Optional[int] = None
+    ) -> 'DNDArray':
+        if n_partitions is None:
+            n_partitions = mt.n_partitions()
+        if block_size is None:
+            block_size = DNDArray.default_block_size
+        if n_partitions == 0:
+            assert mt.count_cols() == 0
+            assert mt.count_rows() == 0
+            t = range_table(0, 0)
+            t = t.annotate(x=0, y=0, block=nd.array([]).reshape((0, 0)))
+            t = t.select_globals(
+                x_field='x',
+                y_field='y',
+                n_rows=0,
+                n_cols=0,
+                n_block_rows=0,
+                n_block_cols=0,
+                block_size=0)
+            return DNDArray(t)
+
+        assert 'y' not in mt.row
+        assert 'x' not in mt.row
+        assert 'block' not in mt.row
+
+        n_rows, n_cols = mt.count()
+        n_block_rows = (n_rows + block_size - 1) // block_size
+        n_block_cols = (n_cols + block_size - 1) // block_size
+        entries, cols, row_index, col_blocks = (Env.get_uid() for _ in range(4))
+        mt = (mt
+              .select_globals()
+              .select_rows()
+              .select_cols()
+              .add_row_index(row_index)
+              .localize_entries(entries, cols))
+        # FIXME: remove when ndarray support structs
+        mt = mt.annotate(**{entries: mt[entries][entry_field]})
+        mt = mt.annotate(
+            **{col_blocks: hl.range(n_block_cols).map(
+                lambda y: hl.struct(
+                    y=y,
+                    entries=mt[entries][(y * block_size):((y + 1) * block_size)]))}
+        )
+        mt = mt.explode(col_blocks)
+        mt = mt.select(row_index, **mt[col_blocks])
+        mt = mt.annotate(x=hl.int(mt[row_index] // block_size))
+        mt = mt.key_by(mt.x, mt.y)
+        mt = mt.group_by(mt.x, mt.y).aggregate(
+            entries=hl.sorted(
+                hl.agg.collect(hl.struct(row_index=mt[row_index], entries=mt.entries)),
+                key=lambda x: x.row_index
+            ).map(lambda x: x.entries))
+        mt = mt.select(block=hl.nd.array(mt.entries))
+        mt = mt.select_globals(
+            x_field='x',
+            y_field='y',
+            n_rows=n_rows,
+            n_cols=n_cols,
+            n_block_rows=n_block_rows,
+            n_block_cols=n_block_cols,
+            block_size=block_size)
+        fname = new_temp_file()
+        mt = mt.key_by(mt.x, mt.y)
+        mt.write(fname, _codec_spec=DNDArray.fast_codec_spec)
+        t = hl.read_table(fname, _intervals=[
+            hl.Interval(hl.Struct(x=x, y=y),
+                        hl.Struct(x=x, y=y + 1))
+            for x in range(n_block_rows)
+            for y in range(n_block_cols)])
```

Comment:
I think we should just use names like `block_row` and `block_col` instead of `x` and `y`. It's too easy to make mistakes with `x` and `y` in my experience. In particular, there's the conflict between `x` seeming like the first axis over which you should vary, while also seeming like it means "moving horizontally", which it doesn't here.